lib/spack/spack/database.py

```diff
@@ -1114,8 +1114,10 @@ def _add(
             new_spec._hash = key
 
         else:
-            # If it is already there, mark it as installed.
+            # If it is already there, mark it as installed and update
+            # installation time
             self._data[key].installed = True
+            self._data[key].installation_time = _now()
```

Comment:
I don't think that will work either. If you record the timestamp before installing and the install process changes anything about the install directory (which it probably will, unless you run `make clean` afterwards), then the user would automatically reinstall it the next time they ran `spack install`, and the next time, and the next, etc.

The case you're worried about is going to be tricky... if the file changed during install, it's actually going to be really hard to know whether that change was included in the previous install. It may have changed just in time to be included, or not in time. I don't know if we can do anything better for that than to offer the `--overwrite` option or instruct users to `touch` a file after the install if they edit the code during the install.

Hash checking is more robust, but it's also much slower. I think timestamps should be sufficient. The only case I can think of that the timestamp won't update properly is if someone changes the timestamp on the file manually. Is there another case you're thinking of that I should be worried about? I already check the directory timestamps in addition to the file timestamps to make sure I account for deleted files.

Reply:
> I don't think that will work either. If you record the timestamp before installing and the install process changes anything about the install directory

Initially I thought of copying the contents to another location, but this isn't always viable (e.g. if the build has hardcoded its prefix somewhere). We could probably figure out which types of files to exclude when creating the timestamp (e.g. ignore things like `.o` files), but I don't think that makes sense to do in this PR.

> The only case I can think of that the timestamp won't update properly is if someone changes the timestamp on the file manually.

System time synchronization issues (e.g. clock adjustments) may also cause transient problems (although I'd expect that to be rare).

Overall it isn't essential I think to come up with the perfect change detection logic, but it may need revisiting in a later PR.

Reply:
> it may need revisiting in a later PR.

I agree. I think we'll get a sense for what users expect in this particular corner case as it goes into use too.

Reply:
One potential optimization you could make is when you rerun the build, check the latest timestamp to see if anything in the build directory changed at all. If it doesn't you could skip the installation stage which would save some time at least for larger packages with complex installers.